Replace the Remaining Parts of the FORTRAN Initialization Procedure

davschneller opened the pull request:
This PR shall replace all remaining dependencies on the FORTRAN code with C++ code. This means that:

* The parameter reading (from `readpar.f90`) was translated to C++ structs. This is done by using the YAML structure from the `f2y` package, and creating structs from there, just like for the dynamic rupture or the LTS parameters. See `InputParameters.hpp` for the result. The reader also warns for old, unused parameters which were still allowed by the FORTRAN namelists. The whole thing is far from perfect, but it works (until we implement a new input file format from scratch, I'd guess). Right now, it is not merged with the DR and LTS parameter readers, and the DR-related namelist sections will throw warnings in the log (which we can still change).
* Source file reading for format 50 (FSRM). Includes some tests for the new reader. The previous reader was located in the `readpar.f90` file.
* Attenuation fitting for visco-elastic materials. This also includes some small unit tests. The previous attenuation fitting was located in `ini_model.f90`.
* The `LtsLayout` class now uses the same timestepping functions as `LtsWeights`. Before, it still used the results from the FORTRAN code which in essence computed the same function. The parameter DB was also adjusted to use material averaging (and not only barycenters) in the case of PUML by generalizing the element generator classes. The timestepping was done mainly in `dg_setup.f90` and `calc_deltat.f90` before.
* The model parameters for the LTS tree are now set in C++. For now, we compute the ghost layer materials on each rank instead of transferring it (this can be changed again to sending the materials via MPI, however). Previously, this was mainly done in `dg_setup.f90` and `mpiexchangevalues.f90`.
* The call structure (mainly from `dg_setup.f90`) was translated to C++, since after the previous points, we are left with only a FORTRAN skeleton which calls C++ methods. For that, see a follow-up post with all the details, if you want to check them. See `Initializer/InitProcedure` which borrows a lot from the functions in the `Interoperability` class. Also, the mesh reading does now not need to translate the mesh data back to FORTRAN for the timestep calculation anymore.
* Removed all FORTRAN code and the `Interoperability` class.
* Updated the documentation and removed all deprecated parameters and sections from it.
* Removed the legacy Gambit3D (fast) reader.
* The `Geometry` module was mostly put into its own namespace under `seissol::geometry` (excluding `MeshDefinition.h`). Also, the code from `NetcdfReader` and `MeshReader` was mostly moved to cpp files.

Apart from that, there should be no more functionality left in the FORTRAN code. See https://github.com/SeisSol/SeisSol/pull/829#issuecomment-1497572994 for a more detailed analysis on what got replaced.
The resulting C++ structure are most likely not architecturally completely thought through, since this PR focused only on removing the FORTRAN part—giving the initialization procedure a proper and clean structure should however be more something of PRs which are still to come.

The following open questions/issues remain for this PR, or for the future:

* The translated attenuation code does not seem to support multithreading right now (maybe due to the Eigen solve against a pseudo-inverse?)
* Where to put the attenuation fitting and the model parameter setting in general? It would probably be best if it was included in the material class initialization or called by the parameter DB right after the material has been averaged. Or for that matter, maybe one could merge even more from `initializeCellMaterial` into the parameter DBs.
* How should we handle the ghost region cell materials? We need those for the Riemann problems. In the FORTRAN code, the materials were transferred using MPI (after being serialized into a `real` array, and serialized back afterwards). Right now, the new implementation re-computes the ghost region materials, for which we only need the vertices and material groups. However, we could go back to transferring the material classes instead.
* The helper struct `CelltoVertexArray` (for element generators) may be transformed into a more object-oriented form (i.e. subclass from it, instead of supplying function pointers).
* Merge the DR and LTS parameter reading with the main parameter reader (and maybe eventually implement a new input file format from scratch... It would be also nice to have something like Rust enums in C++...).

Thomas-Ulrich commented:
Hi, 
I suggest already applying clang-format to the code, or else the referenced lines of our reviews will be messed up when you do this step.

davschneller commented:
That may be a good idea... Thanks! It's done now.

davschneller commented:
As sort of promised in the original PR description, here's a more detailed analysis of the remaining FORTRAN initialization code—and where the code ended up afterwards—in case someone wants to track the changes a bit better. Otherwise, feel free to ignore this message. :)

We begin in `seissolxx.f90`. However, most of the initialization itself happens in `dg_setup.f90` (referenced from `ini_seissol.f90`).
We'll mostly look at the C interop calls in the following, and how they were translated (mostly copy-past of the interop methods). I'll reference the header files here, but there should always be an associated source file. Any FORTRAN method which is not mentioned here is (probably) not relevant, or already obsolete now.

* `readpar`: reads the parameter file (re-written in `InputParameters.hpp`, but it right now ignores DR and most LTS parameters). It also contains a reader for source format 50 (which has been translated, cf. `FSRMSource.hpp` etc.)
* if checkpointing is enabled: `c_interoperability_enableCheckPointing` (moved to `InitIO.hpp`)
* if output is enabled: `c_interoperability_enableWaveFieldOutput` (moved to `InitIO.hpp`)
* `c_interoperability_getIntegrationMask` (moved to `InitIO.hpp`)
* `read_mesh_fast`: read the mesh. This, and the contained methods have been moved to `InitMesh.hpp`.
    * `read_mesh_{puml,netcdf,gambitfast}_c`:
        * call read_mesh evenually which translates the mesh back to FORTRAN. This has been removed for `InitMesh.hpp`.
        * contain a hook for `POST_MESH`. The hook has been copied, and extended by a `PRE_MESH` hook which is run before the whole procedure—in case it should ever be needed.
        * in FORTRAN, some more ops are done which are used for the timestep calculation later-on. This has also been removed.
    * `read_mesh_puml_c` also invokes LtsWeights, the other two do not do this. This did not change during the rewrite.
* `iniGalerkin3D_us_level1_new`: initializes just some parameters. It became obsolete during the translation process.
* `ini_OptionalFields`: allocates the material space and timestep parameters in FORTRAN. It also became obsolete.
* `ini_MODEL`:
    * pre-model hook. Cf. `InitModel.hpp`.
    * `c_interoperability_initializeModel`: fetches model parameters per element; obtains maximum wave speed (duplicate to LtsWeights?). This part of the procedure, i.e. querying the model, has been moved to `InitModel.hpp` (method `initializeCellMaterial`, towards the beginning of the method). However, we do not copy the model to a `real` array anymore, but initialize the cells directly.
    * (*) for all elements: `fitAttenuation`: needed for visco-elastic models (and only enabled in this case) Moved to `Attenuation.hpp` (in the `Physics` folder right now, feel free to place it closer to the visco-elastic model). Right now, the attenuation fitting is also called in `InitModel.hpp` (method `initializeCellMaterial`).
    * post-model hook. Cf. `InitModel.hpp`.
* `iniGalerkin3D_us_level2_new`:
    * `iniGalerkin3D_us_intern_new`: does nothing apparent anymore, and is therefore obsolete.
    * `BuildSpecialDGGeometry3D_new`: computes insphere radius, for timestepping (depends on the mesh info gathered in `read_mesh_fast` before). This method also got removed during translation, as it was replaced by the timestepping code from `LtsWeights`.
    * `c_interoperability_setInitialConditionType`. Has been moved to `InitSideConditions.hpp`. It does not need the LTS to not exist, therefore it could be moved back in the initialization procedure, without any apparent effect.
    * `c_interoperability_setTravellingWaveInformation`. As above, see `InitSideConditions.hpp`.
    * `ini_calc_deltaT`: initialization for timestep calculation. Got removed because obsolete.
    * `cfl_step`: calculates the timestep. The formula is exactly identical to the LtsWeights formula, and therefore the translated code uses the `LtsWeights` code instead, cf. `GlobalTimestep.hpp`. We also support averaged materials for timestepping now (to match the actual material on the cell).
    * for all elements: `c_interoperability_setTimeStepWidth`: sets the timestep in the LtsLayout. This became obsolete; now the timestep is set once `LtsLayout` gets its `MeshReader` set.
    * `c_interoperability_initializeClusteredLts`: sets up timestepping (calling LtsLayout a bunch of times), creates the Luts. This code has been moved without much alterations to `InitModel.hpp`. Only the `Lut` was made a member of the `MemoryManager`, since it seemed to be equally used to the `LTSTree` etc.
    * `MPIExchangeBackground`: exchanges ghost and copy materials (the `real` arrays from the `ini_MODEL` procedure). This has for now been replaced by computing the ghost cell materials locally—at a higher init cost. It could again be replaced with a transfer, at the cost of a higher code complexity, I guess.
    * for the element, and all sides: `c_interoperability_setMaterial`. Sets the material for a cell, and the neighbor cell material for all sides. Thus, we also require the materials from the ghost cells—which is why we did the earlier step. This is done in `InitModel.hpp`, in the method `initializeCellMaterial`, and there, in the loop over the tree.
    * `c_interoperability_synchronizeCellLocalData`: copied, and moved to `InitModel.hpp`.
    * `c_interoperability_initializeMemoryLayout`: copied, and moved to `InitModel.hpp`.
    * `c_interoperability_initFaultOutputManager`: copied, and moved to `InitIO.hpp`.
    * source init: `c_interoperability_setupNRFPointSources`, `c_interoperability_setupFSRMPointSources`. Moved to `InitSideConditions.hpp`, and `SourceManager.hpp`.
    * `c_interoperability_initializeEasiBoundaries`: copied, moved to `InitSideConditions.hpp`.
    * `c_interoperability_initializeGravitationalAcceleration`: copied to `Init.hpp`, and it is now called before the model is even read.
    * `c_interoperability_initializeCellLocalMatrices`: copied, moved to `InitModel.hpp`.
* `icGalerkin3D_us_new`: project initial conditions, set some material parameters
    * `c_interoperability_projectInitialField`: copied, moved to `InitSideConditions.hpp` and merged with the other initial condition-related methods.
    * if plasticity is enabled: (all of this has been copied to `InitModel.hpp` and integrated with the model parameter setting procedure, i.e. method `initializeCellMaterial`, LTStree loop)
        * for all elements: `c_interoperability_setInitialLoading`, `c_interoperability_setPlasticParameters `
        * `c_interoperability_setTv`: done after the LTStree loop in `initializeCellMaterial`, `InitModel.hpp`.
        * `c_interoperability_synchronizeCellLocalData`. Copied to `InitModel.hpp`. Since we set the plasticity and the material parameters once, we only need such a synchronization once.
* `inioutput_SeisSol`: All methods here have been copied to `InitIO.hpp`.
    * `c_interoperability_enableFreeSurfaceOutput`
    * `c_interoperability_initializeIO`
* `calc_SeisSol`: see `Init.hpp`.
    * `c_interoperability_report_device_memory_status`.
    * `c_interoperability_simulate`: runs the simulation.
* `close_SeisSol`: see `Init.hpp`.

And that's it. I've tested the setup by now, and for now the energy terminal output yields the same results as before. (still, I'm probably going to do a few more tests)

ravil-mobile commented:
Hi @davschneller, many thanks for removing old Fortran code! We're still waiting for LRZ to fix issues with the could. Once it is done, we can trigger all our pipelines and test whether SeisSol works as expected with those changes.

Thomas-Ulrich commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+  ssp.model.plasticity = reader.readWithDefault("plasticity", false);
+  ssp.model.tv = reader.readWithDefault("tv", 0.1);
+  ssp.model.useCellHomogenizedMaterial =
+      reader.readWithDefault("usecellhomogenizedmaterial", false);
```
default is true
https://github.com/SeisSol/SeisSol/blob/master/src/Reader/readpar.f90#L281

Thomas-Ulrich reviewed (changes requested):
Nice code, see detailed comments below.

krenzland commented on src/SourceTerm/Manager.cpp:
```diff
@@ -255,54 +260,61 @@ void seissol::sourceterm::Manager::mapPointSourcesToClusters(const unsigned* mes
+                                               seissol::initializers::Lut* ltsLut,
+                                               time_stepping::TimeManager& timeManager) {
+  if (sourceType == SourceType::NrfSource) {
+    logInfo(seissol::MPI::mpi.rank()) << "Reading an NRF source (type 42).";
```
an NRF is correct because "N" starts with a vowel

Thomas-Ulrich replied:
Ha ok, I got it now.

krenzland commented on src/SourceTerm/Manager.cpp:
```diff
@@ -255,54 +260,61 @@ void seissol::sourceterm::Manager::mapPointSourcesToClusters(const unsigned* mes
+                  "library. However, this is not the case for this build.";
+#endif
+  } else if (sourceType == SourceType::FsrmSource) {
+    logInfo(seissol::MPI::mpi.rank()) << "Reading an FSRM source (type 50).";
```
same as above

krenzland commented on src/SourceTerm/Manager.cpp:
```diff
@@ -71,28 +73,28 @@
-        return value[i] < value[j];
-    }
+template <typename T>
+class index_sort_by_value {
```
indexSortByValue?

davschneller replied:
Replaced by lambda

krenzland reviewed (changes requested):
Looks mostly good! Often, you do not include functions in the header file

krenzland reviewed (changes requested):
Looks mostly good! Often, you do not include functions in the header file

davschneller commented on src/Physics/Attenuation.cpp:
```diff
@@ -0,0 +1,102 @@
+    }
+  }
+
+  auto qpinv = Eigen::VectorXd::Constant(kmax, 1 / vm.Qp);
+  auto qsinv = Eigen::VectorXd::Constant(kmax, 1 / vm.Qs);
```
I know, happened to me already during writing all of this... We should be fine here, but you're right, I'll change it for clarity.

davschneller replied:
Changed.

davschneller commented on src/Geometry/MeshReader.h:
```diff
@@ -375,6 +412,69 @@ class MeshReader
+		std::vector<MPI_Status> sendStatus{numMPIDomains};
+		std::vector<MPI_Status> recvStatus{numMPIDomains};
+
+		MPI_Waitall(numMPIDomains, &sendRequests[0], &sendStatus[0]);
+		MPI_Waitall(numMPIDomains, &recvRequests[0], &recvStatus[0]);
```
For the group transfer, it's basically the code from the method above it for the vertex transfer (the two methods could be merged, maybe); i.e. just copied and adjusted legacy code. :) But ok, it can be re-written to be a bit prettier.

davschneller commented on src/Initializer/InitProcedure/InitModel.cpp:
```diff
@@ -0,0 +1,399 @@
+    for (unsigned dup = 1; dup < seissol::initializers::Lut::MaxDuplicates &&
+                           meshToLts[dup][meshId] != std::numeric_limits<unsigned>::max();
+         ++dup) {
+      memcpy(reinterpret_cast<void*>(&var[meshToLts[dup][meshId]]),
+             reinterpret_cast<void*>(ref),
+             sizeof(T));
```
Agreed, I'll look into it. So far, it's legacy code, just copied over and continuing to do its job as before.

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+    for (const auto& subnodes : node) {
+      auto field = subnodes.first.as<std::string>();
+      if (visited.find(field) == visited.end()) {
+        logWarning(seissol::MPI::mpi.rank()) << "The field" << field << "is not known to SeisSol.";
```
Do you mean: all fields that were read with default value?

Also, we should be able to ignore the sections which belong to the DR part and the LTS parameters in the reader (i.e. `markUnused`).

davschneller replied:
Ok, done—they are marked as unused for now. We can discuss merging the whole parameter parsing machinery preferrably after this PR, I'd say.

davschneller commented on src/Initializer/InitProcedure/Init.hpp:
```diff
@@ -0,0 +1,13 @@
+
+namespace seissol::initializer::initprocedure {
+
+void seissolMain();
```
I didn't do it, because they are not needed outside of the cpp file. So yes, they should probably be marked as `static`. I'll change that.

davschneller commented on src/Geometry/MeshReader.h:
```diff
@@ -52,6 +52,12 @@
 #include "Parallel/MPI.h"
 
+enum class MeshFormat {
+	Gambit3D,
```
As Thomas mentioned, we could retire the format just now. The reader is still being compiled right now, but I haven't tested it.

davschneller replied:
It's gone.

davschneller commented on src/Initializer/InitProcedure/InitMesh.cpp:
```diff
@@ -0,0 +1,175 @@
+  switch (meshformat) {
+  case MeshFormat::Gambit3D:
+    seissol::SeisSol::main.setMeshReader(
+        new GambitReader(commrank, ssp.mesh.meshFileName.c_str(), partitionfile.c_str()));
```
Ok, then it shall be gone. :) Should we still leave some message when loading a Gambit mesh like "the format is not supported anymore, please convert your files to PUML"—or just fail with "unknown mesh type" or similar?

davschneller replied:
It's been deleted.

Thomas-Ulrich replied:
fail with "unknown mesh type"

davschneller commented on src/SourceTerm/FSRMReader.cpp:
```diff
@@ -0,0 +1,147 @@
+ * @section DESCRIPTION
+ **/
+
+#include "FSRMReader.h"
```
It's "Finite Source Rupture Model"—that's how it was called in the FORTRAN code, if I remember correctly

krenzland commented on src/Geometry/MeshReader.h:
```diff
@@ -144,6 +158,29 @@ class MeshReader
+      double x = m_vertices[vertexNo].coords[0];
+      double y = m_vertices[vertexNo].coords[1];
+      double z = m_vertices[vertexNo].coords[2];
+      for (unsigned i = 0; i < 3; ++i) {
+        m_vertices[vertexNo].coords[i] = scalingMatrix[0][i] * x + scalingMatrix[1][i] * y + scalingMatrix[2][i] * z;
+      }
```
Scaling makes sense (e.g. to switch to a coordinate system where wavespeeds are in km/s or similar), but rotations etc. are a bit weird. I'm not sure when this would be useful. @Thomas-Ulrich any idea?

Thomas-Ulrich replied:
I never used that and I don't know anybody who used it. But probably it was useful at some point if somebody implemented it :-) Any idea @AliceGabriel ?

davschneller commented:
I'll collect the open issues from the reviews above (some commits fixing some things may come soon):

* If we really want to remove support for Gambit... What about `preprocessing/meshing/gambit2netcdf`? That's the only place where we still have a reference to the `GambitReader` after removing it from all others.
* What to do about the scaling? We could either restrict it (i.e. only displacement and rotations)—or go the other way and maybe add some arbitrary function support? (e.g. using easi, if that's possible) But maybe that's all something for another commit. The main question here is: should these two functions stay, or should we throw them out? If we plan to throw them out but add something better in the future, then a separate feature request issue may be best, I think.

krenzland commented:
_If we really want to remove support for Gambit... What about preprocessing/meshing/gambit2netcdf? That's the only place where we still have a reference to the GambitReader after removing it from all others._
Remove it, not used anymore

_What to do about the scaling? We could either restrict it (i.e. only displacement and rotations)—or go the other way and maybe add some arbitrary function support? (e.g. using easi, if that's possible) But maybe that's all something for another commit. The main question here is: should these two functions stay, or should we throw them out? If we plan to throw them out but add something better in the future, then a separate feature request issue may be best, I think._
A seperate PR would be best IMO. Not sure whether we should throw them out (they are anyway  preserved in the git history?) Keeping them doesn't hurt much though....

davschneller commented on src/Initializer/InitProcedure/InitSideConditions.cpp:
```diff
@@ -0,0 +1,141 @@
+  }
+}
+
+void seissol::initializer::initprocedure::initSideConditions() {
```
It's indeed intended more as a collection of all parts which sort of belong to the model, but not the main waveprop/DR logic. So I had no good name for it. :/

davschneller commented on src/Geometry/MeshReader.h:
```diff
@@ -375,6 +412,69 @@ class MeshReader
+         * */
+	void exchangeGroupsWithMPINeighbors() {
+		// TODO: (if there is time) generalize halo exchange to arbitrary data maybe?
+		size_t numMPIDomains = m_MPINeighbors.size();
```
Completely re-written and combined with the other method out there which also exchanges ghost data.

davschneller commented on src/Equations/datastructures.hpp:
```diff
@@ -19,4 +19,18 @@
 
+namespace seissol::model {
+#if defined(USE_ANISOTROPIC)
+using MaterialClass = AnisotropicMaterial;
```
Sounds a bit C-like... I'm just not so much a fan of underscores, but we could go with it. Should we?

ravil-mobile replied:
Maybe it is better to name it like `MaterialType`?

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+  logInfo(seissol::MPI::mpi.rank()) << "SeisSol parameter file read successfully.";
+}
+
+void SeisSolParameters::printInfo() {
```
Removed.

davschneller commented on src/Initializer/BasicTypedefs.hpp:
```diff
@@ -50,12 +50,23 @@ enum class FaceType {
+// Once the FaceType enum is updated, make sure to update these methods here as well.
+
+// Checks if a face type is an interior face (i.e. there are two cells adjacent to it).
+constexpr bool isInteriorFaceType(FaceType faceType) {
```
The functionality is slightly different however, namely that the periodic faces are classified as interior (since plain, DR, and periodic are often treated as interior faces throughout the code)

davschneller commented on src/Geometry/MeshReader.h:
```diff
@@ -144,6 +158,29 @@ class MeshReader
   // scalingMatrix is stored column-major, i.e.
   // scalingMatrix_ij = scalingMatrix[j][i]
   void scaleMesh(double const scalingMatrix[3][3])
```
Relic from removing FORTRAN. :) Gone now.

davschneller commented on src/Initializer/InitProcedure/InitMesh.cpp:
```diff
@@ -0,0 +1,175 @@
+    VrtxCoords center{drParameters->referencePoint[0],
+                      drParameters->referencePoint[1],
+                      drParameters->referencePoint[2]};
+    meshReader.findFault(center, drParameters->refPointMethod);
```
`extractFaultInformation` sounds good; I've re-named it.

davschneller commented on src/Initializer/InitProcedure/InitModel.cpp:
```diff
@@ -0,0 +1,399 @@
+  for (const auto& neighbor : meshReader.getMPINeighborVertices()) {
+    ghostIdxMap[neighbor.first].reserve(neighbor.second.size());
+    for (const auto& vertices : neighbor.second) {
+      ghostIdxMap[neighbor.first].push_back(ghostVertices.size());
```
We need to "linearize" the unordered map, so we can compute the materials for all ghost cells. Thus, we need to go back to an array, unfortunately. Or we implement the C2VArray for a unordered_map somehow

davschneller commented on src/Initializer/time_stepping/GlobalTimestep.cpp:
```diff
@@ -0,0 +1,69 @@
+        for (unsigned cell = 0; cell < ctov.size; ++cell) {
+            double pWaveVel = materials[cell].getMaxWaveSpeed();
+
+            // Compute insphere radius
```
Done. Should we make it available in the header—or keep it `static`?

Thomas-Ulrich replied:
@krenzland ?

davschneller commented on src/Initializer/time_stepping/GlobalTimestep.cpp:
```diff
@@ -0,0 +1,69 @@
+            double insphere = std::fabs(alpha) / (Nabc + Nabd + Nacd + Nbcd);
+
+            // Compute maximum timestep
+            timestep.elementTimeStep[cell] = std::fmin(maximumAllowedTimeStep, cfl * 2.0 * insphere / (pWaveVel * (2 * CONVERGENCE_ORDER - 1)));
```
That is correct—however, we call it with CFL=1 in the place where it was before (i.e. `LtsWeights`). However, this timestep calculation is now also used for the procedure that was in FORTRAN so far, and here the CFL parameter from the parameter file was applied. (`calc_deltat.f90` iirc)

(the timestep was calculated _twice_ so far, one time before re-distributing the mesh (to estimate the computational weight of the cells), and one time for determining the timesteps of the distributed cells (which is used for dividing the cells into time clusters on each rank)—the latter was done in FORTRAN so far)

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+
+  // output: refinement
+  ssp.output.receiverParameters.enabled =
+      reader.readWithDefault("receiveroutput", true); // TODO: document
```
It shouldn't be—the default behavior is `true`, i.e. the values of other parameters (e.g. output interval) count, and the parameter is not expected to be present. It's basically supposed to be a parameter which can disable the receiver output—regardless of what any other parameter says.

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+  ssp.output.loopStatisticsNetcdfOutput =
+      reader.readWithDefault("loopstatisticsnetcdfoutput", false);
+
+  // TODO: check if ioutputmaskmaterial is still in use...
```
Ok, it's gone.

davschneller commented on src/Initializer/InitProcedure/Init.cpp:
```diff
@@ -0,0 +1,80 @@
+  if (device.api->getCurrentlyOccupiedMem() > device.api->getMaxAvailableMem()) {
+    std::stringstream stream;
+
+    stream << "Device(" << rank << ")  memory is overloaded.\n"
+           << "Totally allocated device memory, GB: " << device.api->getCurrentlyOccupiedMem() / GB
```
@ravil-mobile

davschneller commented on src/Initializer/InitProcedure/InitIO.cpp:
```diff
@@ -0,0 +1,222 @@
+                      ssp.output.energyParameters.terminalOutput,
+                      ssp.output.energyParameters.computeVolumeEnergiesEveryOutput,
+                      ssp.output.prefix.c_str(),
+                      ssp.output.energyParameters.interval);
```
I've put in `ssp.output.energyParameters` for now; seems cleaner to me. But I can also forward all output parameters instead.

davschneller commented on src/Initializer/InitProcedure/InitIO.cpp:
```diff
@@ -0,0 +1,222 @@
+                        ssp.output.prefix,
+                        ssp.output.receiverParameters.interval,
+                        ssp.output.receiverParameters.samplingInterval,
+                        ssp.output.receiverParameters.computeRotation);
```
Passed `ssp.output.receiverParameters` for now

davschneller commented on src/Initializer/InitProcedure/InitSideConditions.cpp:
```diff
@@ -0,0 +1,141 @@
+  auto& memmng = seissol::SeisSol::main.getMemoryManager();
+  std::vector<std::unique_ptr<physics::InitialField>> iniconds;
+  std::string initialConditionDescription = "";
+  if (icparams.type == seissol::initializer::parameters::InitializationType::Planarwave) {
```
I don't quite like the sematics of switches in C/C++ to be honest—so I'd leave it a large `if` for now

davschneller commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,196 @@
+  Travelling,
+  Scholte,
+  Snell,
+  Ocean0,
+  Ocean1,
+  Ocean2
```
Thing is, the old parameter was a string, and the new one is an enum. And we don't have the variadic type enums like in e.g. Rust, unfortunately.

davschneller commented on src/Initializer/InitProcedure/InitIO.cpp:
```diff
@@ -0,0 +1,222 @@
+        ssp.output.waveFieldParameters.plasticityMask,
+        ssp.output.waveFieldParameters.bounds,
+        ssp.output.waveFieldParameters.groups,
+        ssp.output.xdmfWriterBackend);
```
Passed `ssp.waveFieldParameters`

Thomas-Ulrich commented on src/Initializer/InitProcedure/InitModel.cpp:
```diff
@@ -0,0 +1,399 @@
+  initializers::MemoryManager& memoryManager = seissol::SeisSol::main.getMemoryManager();
+  unsigned* const(&meshToLts)[seissol::initializers::Lut::MaxDuplicates] =
+      memoryManager.getLtsLut()->getMeshToLtsLut(handle.mask);
+  unsigned* duplicatedMeshIds = memoryManager.getLtsLut()->getDuplicatedMeshIds(handle.mask);
```
thank you

Thomas-Ulrich commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,196 @@
+
+struct OutputEnergyParameters : public OutputBaseParameters {
+  bool terminalOutput = false;
+  bool computeVolumeEnergiesEveryOutput = true;
```
is not resolved

Thomas-Ulrich commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+          {{"none", seissol::checkpoint::Backend::DISABLED},
+           {"posix", seissol::checkpoint::Backend::POSIX},
+           {"hdf5", seissol::checkpoint::Backend::HDF5},
+           {"mpio", seissol::checkpoint::Backend::MPIO},
+           {"mpio_async", seissol::checkpoint::Backend::MPIO_ASYNC},
+           {"sionlib", seissol::checkpoint::Backend::SIONLIB}});
```
There are unit test in the checkpointing submodules. I think when checkpointing is fixed, it should be easy to make it work for all backends (but is it useful? probably not).

davschneller commented:
The last commit 9aa66ac12bd76809227a92f135595c9389c3d611 may need special attention, since it increases the limit for the timestep validation for the fault receivers a bit (which are oriented at the cluster timesteps, if I've seen that correctly... So maybe something _did_ change here when either switching away from FORTRAN or it's just some bug somewhere in the new code). I've tested for `tpv5/double`, and the difference was at around 1.4e-14 which is slightly above the threshold of 1e-14. For now, it's increased to 1e-12.

davschneller commented on src/Initializer/InitProcedure/InitModel.cpp:
```diff
@@ -0,0 +1,399 @@
+template <typename T>
+void synchronize(const seissol::initializers::Variable<T>& handle) {
+  initializers::MemoryManager& memoryManager = seissol::SeisSol::main.getMemoryManager();
+  unsigned* const(&meshToLts)[seissol::initializers::Lut::MaxDuplicates] =
```
Thus it has been become an `auto*`.

davschneller commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,196 @@
+#endif
+}
+
+struct ModelParameters {
```
It shouldn't matter—but you got a point. I've removed them.

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+
+  void markUnused(const std::string& field) {
+    logDebug(seissol::MPI::mpi.rank())
+        << "The field" << field << "is ignored (regardless of if it exists or not)";
```
It means that the field will be ignored during the `warnUnused` procedure.

davschneller commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,196 @@
+  // TODO: port rest of the DR parameters here?
+};
+
+enum class OutputFormat { None = 10, Xdmf = 6 };
```
There is now `WavefieldOutput` as a field (and also a `Checkpoint` field, working as a guard for checkpointing). Will be added to the docu.

We could retire `OutputFormat`—however, this would most likely constitute a breaking change. I've put in a notice that it'll still work, but it is better to use `WavefieldOutput`, and the `outputformat` parameter may be removed in the future (most likely when we introduce a new format anyways).

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+  // output: energy
+  ssp.output.energyParameters.enabled = reader.readWithDefault("energyoutput", false);
+  ssp.output.energyParameters.interval = reader.readWithDefault("energyoutputinterval", 1.0e100);
+  ssp.output.energyParameters.enabled &= ssp.output.energyParameters.interval > 0;
```
Yes, if we enter `energyoutputinterval = 0` in our parameter file. Not a nice way to disable it, but it works. I've added an info/a warning to tell the user to set `energyoutput = 0` instead.

davschneller commented on src/Initializer/InputParameters.cpp:
```diff
@@ -0,0 +1,462 @@
+      reader.readWithDefault("receiveroutput", true); // TODO: document
+  ssp.output.receiverParameters.interval =
+      reader.readWithDefault("receiveroutputinterval", 1.0e100);
+  ssp.output.receiverParameters.enabled &= ssp.output.receiverParameters.interval > 0;
```
Cf. https://github.com/SeisSol/SeisSol/pull/829#discussion_r1185214954

davschneller commented:
A huge thank you to everyone who has already taken a look at this a bit-too-massive PR. All the addressed issues should be handled by now (I hope).

Once https://github.com/SeisSol/Examples/pull/62 is merged and the CPU runner is fixed, the tests should be green as well.

ravil-mobile commented:
> A huge thank you to everyone who has already taken a look at this a bit-too-massive PR. All the addressed issues should be handled by now (I hope).
> 
> Once [SeisSol/Examples#62](https://github.com/SeisSol/Examples/pull/62) is merged and the CPU runner is fixed, the tests should be green as well.

Hi, the GPU pipeline will be fixed once you merge this branch with the master.

ravil-mobile reviewed (commented):
Could you remove this line while you will be doing a merge with the master?

ravil-mobile commented on src/sources.cmake:
```diff
@@ -76,34 +81,10 @@ src/ResultWriter/FreeSurfaceWriter.cpp
 src/ResultWriter/EnergyOutput.cpp
 
 # Fortran:
```
Could you, please, delete this line while you will be doing a merge with the master?

ravil-mobile commented on src/Initializer/InitProcedure/InitIO.cpp:
```diff
@@ -0,0 +1,206 @@
+#include "Parallel/MPI.h"
+
+static void setupCheckpointing() {
+  const auto& ssp = seissol::SeisSol::main.getSeisSolParameters();
+  auto& memmng = seissol::SeisSol::main.getMemoryManager();
```
I wouldn't use names like `ssp` and `memmng`. It is difficult to read and get semantic out of it. I would use `seissolParameters` (or `seissolParams`) and `memoryManager`.

I can see that it is in many places in the code. I just highlighted only here.

ravil-mobile commented:
Hi, @davschneller. Thanks for changes. The CI pipelines didn't fail which is a good sign! 

There is one which will make this PR better. So, there are many small and dedicated commits. Maybe you can work with `git squash`  to squash some small commits into a single one. Don't try to squash `merge` commits but rather focus on the ones which are in between.

I know that it is annoying but the commit history will look better.

What do you think?

ravil-mobile commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,214 @@
+  MeshParameters mesh;
+  InitializationParameters initialization;
+  OutputParameters output;
+  TimesteppingParameters timestepping;
```
`timestepping` -> `timeStepping`

ravil-mobile replied:
`timestepping` - > `timeStepping`

ravil-mobile commented on src/Initializer/InputParameters.hpp:
```diff
@@ -0,0 +1,214 @@
+  SourceParameters source;
+  EndParameters end;
+
+  void readPar(const YAML::Node& baseNode);
```
I would write `readParameters` instead of `readPar`

Thomas-Ulrich commented on src/Initializer/time_stepping/LtsLayout.cpp:
```diff
@@ -73,27 +78,26 @@ seissol::initializers::time_stepping::LtsLayout::~LtsLayout() {
+  auto timesteps = seissol::initializer::computeTimesteps(seissolParams.timestepping.cfl, seissolParams.timestepping.maxTimestepWidth, seissolParams.model.materialFileName,
+        seissol::initializers::CellToVertexArray::fromMeshReader(i_mesh));
+  
+  m_cellTimeStepWidths = new double[m_cells.size()];
```
is there a good reason not to use a std::vector?

davschneller replied:
There is actually none to not use it... So it's done now. :) Refactoring more of `LTSLayout` will stay a task for the future though, I'd guess.

Thomas-Ulrich commented on src/Initializer/InitProcedure/InitMesh.cpp:
```diff
@@ -0,0 +1,180 @@
+#if USE_NETCDF
+    realMeshFileName = seissolParams.mesh.meshFileName + ".nc";
+    logInfo(seissol::MPI::mpi.rank())
+        << "By old SeisSol conventions for Netcdf meshes, the Netcdf file extension \".nc\" is "
```
I would just write:
Netcdf file extension \".nc\" appended. Updated mesh filename:

Thomas-Ulrich commented on src/Geometry/PUMLReader.cpp:
```diff
@@ -59,207 +59,191 @@
+    puml.open((file + ":/connect").c_str(), (file + ":/geometry").c_str());
+    puml.addData((file + ":/group").c_str(), PUML::CELL);
+    puml.addData((file + ":/boundary").c_str(), PUML::CELL);
         const auto numTotalCells = puml.numTotalCells();
```
- a lot of formating changes, but not consistent with the unchanged lines, it seems.
At this point I would consider applying clang-format.

davschneller commented on src/ResultWriter/WaveFieldWriter.h:
```diff
@@ -72,183 +76,180 @@ namespace writer
-            int* plasticityMask,
-            const double* outputRegionBounds,
-            const std::unordered_set<int>& outputGroups,
+			const seissol::initializer::parameters::WaveFieldOutputParameters& parameters,
```
I hadn't applied on this file (and for some reason just decided to tabs->spaces...), but now it's in the whitelist.

davschneller commented on .ci/format.sh:
```diff
@@ -3,12 +3,40 @@
 # sample command line usage: $0 clang-format(version >= 6.0) $SEISSOL_SOURCE_DIR
-# $ sh ./.dev/format.sh `which clang-format` ./
+# $ sh ./.ci/format.sh `which clang-format` ./
 # $ sh format.sh `which clang-format` ../
```
Ok, fixed.

davschneller commented:
> Hi, @davschneller. Thanks for changes. The CI pipelines didn't fail which is a good sign!
> 
> There is one which will make this PR better. So, there are many small and dedicated commits. Maybe you can work with `git squash` to squash some small commits into a single one. Don't try to squash `merge` commits but rather focus on the ones which are in between.
> 
> I know that it is annoying but the commit history will look better.
> 
> What do you think?

Sounds good; I'll go over it some time soon. At least for some commits that should be possible, I hope

Thomas-Ulrich commented:
Honestly, I m not sure it is really necessary to squash commits. Seems like a lot of work.

krenzland commented:
IMO If the commits are meaningful, don't squash them. If you use 5 commits to try something, the should be rebased in one commit. However, we don't really have any hard rules for this currently.

davschneller commented:
Ok, to be honest... I looked at the commit history, and did not see a lot to squash in a straight-forward way, maybe save for some formatting commits further down the line. Other than that, it's harder to combine them easily, to my estimate at least. And just squashing all commits between merges seems to make them too big in my opinion.

So if it is not a major concern, I'd also rather avoid it on second thought. But thanks for the idea still.

ravil-mobile commented:
> IMO If the commits are meaningful, don't squash them. If you use 5 commits to try something, the should be rebased in one commit. However, we don't really have any hard rules for this currently.

Agree, we don't have any rules for this.

>  Ok, to be honest... I looked at the commit history, and did not see a lot to squash in a straight-forward way, maybe save for some formatting commits further down the line. Other than that, it's harder to combine them easily, to my estimate at least. And just squashing all commits between merges seems to make them too big in my opinion.
So if it is not a major concern, I'd also rather avoid it on second thought. But thanks for the idea still.

Formatting commits are the best candidates. It is just a suggestion to make the PR a bit compact. It is optional. The most important is to preserve numerics.

Thomas-Ulrich reviewed (commented):
From my point of view, the PR can be merged.

ravil-mobile commented:
> From my point of view, the PR can be merged.

Agree
